### ISSUE TITLE
Slow Gradle dependency resolution

### ISSUE THREAD

**Issue by fwcd** [fwcd/KotlinLanguageServer#82]
The language server (unexpectedly) invokes Gradle multiple times and takes longer than expected to resolve the dependencies. Example log:

```
[Info  - 19:53:06] main      Connected to client
[Info  - 19:53:06] client    Adding workspace file:///home/.../Kotlin/KotlinLanguageServer to source path
[Info  - 19:53:06] client    Adding 101 files under /home/.../Kotlin/KotlinLanguageServer to source path
[Info  - 19:53:06] client    Searching for dependencies in workspace root /home/.../Kotlin/KotlinLanguageServer
[Info  - 19:53:06] client    Attempting dependency resolution through CLI...
[Info  - 19:53:06] client    Creating temporary gradle file /tmp/classpath3454683406276641897.gradle
[Info  - 19:53:51] client    Successfully resolved dependencies using Gradle dependencies CLI
[Info  - 19:53:51] client    Attempting dependency resolution through CLI...
[Info  - 19:53:51] client    Creating temporary gradle file /tmp/classpath1518563813544658597.gradle
[Info  - 19:53:51] client    Found gradle at /usr/sbin/gradle
[Info  - 19:53:53] client    Successfully resolved dependencies using Gradle dependencies CLI
[Info  - 19:53:53] client    Attempting dependency resolution through CLI...
[Info  - 19:53:53] client    Creating temporary gradle file /tmp/classpath2782203026256202926.gradle
[Info  - 19:53:55] client    Successfully resolved dependencies using Gradle dependencies CLI
[Info  - 19:53:55] client    Attempting dependency resolution through CLI...
[Info  - 19:53:55] client    Creating temporary gradle file /tmp/classpath6875696104469076848.gradle
[Info  - 19:53:57] client    Successfully resolved dependencies using Gradle dependencies CLI
[Info  - 19:53:57] client    Adding 30 files to class path
```

**Comment by Strum355** [comment]
This appears to have been introduced in one of the commits between 7ca3d6619d4625bf68034e1e94a0b1c54defb388 and 2f6018071bdbeff7a7fc525fc75de85c3234355a as global-symbols-provider branch doesnt have an almost 50 second delay between connecting to client and being ready. global-symbols-provider branch still creates multiple gradle temp files however

**Comment by fwcd** [comment]
Maybe we could use [`git bisect`](https://git-scm.com/docs/git-bisect) to find the exact commit that introduced the bug?

**Comment by Strum355** [comment]
Do we want to count the multiple gradle invocations as "bad"? Need to find the commit that introduced that bug first before we can `git bisect` for the slow temp file creation if yes

**Comment by fwcd** [comment]
Strangely, I could not reproduce the issue with the current `master` branch on macOS (at least with [KotlinQuickStart](https://github.com/fwcd/KotlinQuickStart)).

**Comment by fwcd** [comment]
The newest version v0.1.13 has been released to the extension marketplace and should fix the performance issues. Additionally, it provides more detailed information about every folder for which dependencies are resolved:

```
[Info  - 6:37:47 PM] client    Adding 101 files under d:\git\KotlinLanguageServer to source path
[Info  - 6:37:47 PM] client    Searching for dependencies in workspace root d:\git\KotlinLanguageServer
[Info  - 6:37:48 PM] client    Resolving dependencies for additionalWorkspace through CLI...
[Info  - 6:37:48 PM] client    Creating temporary gradle file C:\Users\fw\AppData\Local\Temp\classpath6205170888282662561.gradle
[Info  - 6:37:51 PM] client    Successfully resolved dependencies using Gradle dependencies CLI
[Info  - 6:37:51 PM] client    Resolving dependencies for additionalWorkspace through CLI...
[Info  - 6:37:51 PM] client    Creating temporary gradle file C:\Users\fw\AppData\Local\Temp\classpath8033091242540791339.gradle
[Info  - 6:37:55 PM] client    Successfully resolved dependencies using Gradle dependencies CLI
[Info  - 6:37:55 PM] client    Resolving dependencies for KotlinLanguageServer through CLI...
[Info  - 6:37:55 PM] client    Creating temporary gradle file C:\Users\fw\AppData\Local\Temp\classpath8115151316574514824.gradle
[Info  - 6:37:59 PM] client    Successfully resolved dependencies using Gradle dependencies CLI
[Info  - 6:37:59 PM] client    Resolving dependencies for additionalWorkspace through CLI...
[Info  - 6:37:59 PM] client    Creating temporary gradle file C:\Users\fw\AppData\Local\Temp\classpath5665796461325486797.gradle
[Info  - 6:38:02 PM] client    Successfully resolved dependencies using Gradle dependencies CLI
[Info  - 6:38:02 PM] client    Adding 29 files to class path
```

Note that `additionalWorkspace` (a small workspace used by the unit tests) has been discovered multiple times. The reason for this is that both Gradle and the Java extension (if present) create copies of the source folder, which happens to contain `additionalWorkspace`, during compilation. Thus, the KotlinLanguageServer ends up finding 4 buildfiles in total, which matches the invocation count in the log:

![image](https://user-images.githubusercontent.com/30873659/52431534-b2bd6480-2b08-11e9-8490-315509aeb0c7.png)

![image](https://user-images.githubusercontent.com/30873659/52431567-c5379e00-2b08-11e9-89bd-8c0b0d6bd1db.png)

![image](https://user-images.githubusercontent.com/30873659/52431598-d54f7d80-2b08-11e9-80c8-c5e89383cc29.png)

![image](https://user-images.githubusercontent.com/30873659/52431700-092aa300-2b09-11e9-855e-6c666b16148e.png)

I am closing the issue since this feature is working as intended. Future optimizations (such as whether a single temporary Gradle file would be sufficient or how to improve Gradle's performance) should be discussed in a separate issue.
